local richterm = require "@rose/batteries/richterm"

const BANNER_LINES = table.freeze {
	"‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ĖĄ  ‚ĖĄ ‚Ėą‚ĖÄ‚ĖÄ‚ĖĄ   ‚ĖĄ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą ‚ĖĄ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ĖĄ",
	"‚Ėą‚Ėą‚Ėą ‚Ėą‚Ėą‚Ėą ‚ĖÄ‚ĖÄ ‚ĖĄ‚Ėą‚ĖĄ ‚ĖĄ‚Ėą ‚Ėą‚Ėą‚Ėą‚ĖĄ‚ĖĄ‚ĖĄ  ‚Ėą‚Ėą‚Ėą‚ĖĄ‚ĖĄ‚ĖĄ‚ĖĄ",
	"‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ĖÄ  ‚Ėą‚ĖĄ ‚ĖÄ ‚ĖĄ‚Ėą‚ĖÄ  ‚ĖÄ‚ĖÄ‚ĖÄ‚Ėą‚Ėą‚Ėą ‚Ėą‚Ėą‚Ėą‚ĖÄ‚ĖÄ‚ĖÄ‚ĖÄ",
	"‚Ėą‚Ėą‚Ėą ‚Ėą‚Ėą‚Ėą ‚ĖÄ‚Ėą‚ĖÄ‚ĖÄ ‚ĖÄ‚ĖÄ   ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ĖÄ ‚ĖÄ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ĖÄ",
}

const BANNER = table.concat(BANNER_LINES, "\n")
const BANNER_INDENT = "                                   "

local header = richterm.combine(richterm.bold, richterm.underline)

local function padStart(str: string, length: number)
	return string.rep(" ", length - string.len(str)) .. str
end

local function padEnd(str: string, length: number)
	return str .. string.rep(" ", length - string.len(str))
end

return table.freeze {
	BANNER = BANNER,
	BANNER_LINES = BANNER_LINES,
	BANNER_INDENT = BANNER_INDENT,

	padStart = padStart,
	padEnd = padEnd,
	header = header,
}
